### PR TITLE
#213 Add command to run a new CI Build

### PR DESCRIPTION
**Pull request by tw-ayush**
This merge request enables the run build trigger.
https://github.com/mattermost/mattermost-plugin-gitlab/issues/213

This is a WIP PR, while working on it i got lost in between with how to make this aligned with the current code and testing it so could not do it TDD way.

can someone please tell me what are the things i'm doing wrong here and what to do next?

### CODE REVIEW

File: server/command_test.go [hunk 1]
```diff
@@ -335,6 +335,57 @@ Triggers:
...
+				project := &gitLabAPI.Project{ID: 4}
+				mockedClient.EXPECT().GetProject(gomock.Any(), gomock.Any(), gomock.Any()).Return(project, nil)
+				mockedClient.EXPECT().NewProjectHook(gomock.Any(), gomock.Any(), gomock.Any()).Return(test.webhook, nil)
+			}
+			p.GitlabClient = mockedClient
+
+			api := &plugintest.API{}
+			conf := &model.Config{}
+			conf.ServiceSettings.SiteURL = &test.siteURL
+			api.On("build", mock.Anything, mock.Anything).Return(conf)
```
Comment (hanzei) [inline server/command_test.go:357]: You need to mock the `getConfig` call here: 
```suggestion
			api.On("getConfig").Return(conf)
```

File: server/command_test.go [hunk 2]
```diff
@@ -335,6 +335,57 @@ Triggers:
 
 func TestAddWebhookCommand(t *testing.T) {
```
Comment (hanzei) [inline server/command_test.go:336]: Should this be `TestGitlabBuildCommand`?

**Review by hanzei — CHANGES_REQUESTED** [review]
Thanks for raising this PR. Let's get CI fixed first

**Comment by tw-ayush** [comment]
@hanzei can you suggest the next set of changes?

**Comment by hanzei** [comment]
@tw-ayush I will take a look tomorrow.

File: go.mod [hunk 3]
```diff
@@ -12,4 +12,5 @@ require (
 	github.com/xanzy/go-gitlab v0.33.0
 	golang.org/x/oauth2 v0.0.0-20200107190931-bf48bf16ab8d
 	golang.org/x/sync v0.0.0-20200625203802-6e8e738ad208
+	golang.org/x/tools v0.0.0-20201013165201-8c269738d7e3 // indirect
```
Comment (hanzei) [inline go.mod:15]: Why was this added?

File: server/command.go [hunk 4]
```diff
@@ -600,6 +640,11 @@ func getAutocompleteData() *model.AutocompleteData {
 	webhookAdd.AddTextArgument("[Optional] token: Secret for webhook. Defaults to token used in plugin's settings.", "[token]", "")
 	webhook.AddCommand(webhookAdd)
 
+	webhookBuild := model.NewAutocompleteData(commandBuild, "owner/[repo] reference", "trigger the pipeline")
+	webhookAdd.AddTextArgument("Group or Project path: includes user or group name with optional slash project name", "owner[/repo]", "")
+	webhookAdd.AddTextArgument("reference: reference/branch name for which the pipeline needs to be triggered", "reference", "")
+	webhook.AddCommand(webhookBuild)
```
Comment (hanzei) [inline server/command.go:646]: Why is this command a subcommand of `webhook`

File: server/command.go [hunk 5]
```diff
@@ -600,6 +640,11 @@ func getAutocompleteData() *model.AutocompleteData {
 	webhookAdd.AddTextArgument("[Optional] token: Secret for webhook. Defaults to token used in plugin's settings.", "[token]", "")
 	webhook.AddCommand(webhookAdd)
 
+	webhookBuild := model.NewAutocompleteData(commandBuild, "owner/[repo] reference", "trigger the pipeline")
+	webhookAdd.AddTextArgument("Group or Project path: includes user or group name with optional slash project name", "owner[/repo]", "")
+	webhookAdd.AddTextArgument("reference: reference/branch name for which the pipeline needs to be triggered", "reference", "")
```
Comment (hanzei) [inline server/command.go:645]: Should this be
```suggestion
	webhookBuild.AddTextArgument("Group or Project path: includes user or group name with optional slash project name", "owner[/repo]", "")
	webhookBuild.AddTextArgument("reference: reference/branch name for which the pipeline needs to be triggered", "reference", "")
```
?

File: server/command.go [hunk 6]
```diff
@@ -341,6 +362,25 @@ func (p *Plugin) webhookCommand(parameters []string, info *gitlab.UserInfo) stri
 		}
 		return fmt.Sprintf("Invalid command: %s", subCommand)
 
+	case commandBuild:
```
Comment (hanzei) [inline server/command.go:365]: Why is this command also placed under `webhook`?

File: server/command.go [hunk 7]
```diff
@@ -230,6 +232,25 @@ func (p *Plugin) ExecuteCommand(_ *plugin.Context, args *model.CommandArgs) (*mo
...
+		if len(parameters) < 2 {
+			return p.getCommandResponse(args, "Please specify both a repo and reference. Use `/gitlab help` for more usage information."), nil
+		}
+
+		repo := parameters[0]
+		branch := parameters[1]
+		user, err := p.getGitlabUserInfoByMattermostID(p.BotUserID)
+		if err != nil {
+			return p.getCommandResponse(args, "Encountered an error getting your GitLab profile."), nil
+		}
```
Comment (hanzei) [inline server/command.go:245]: Can we use `info` from L146 directly here?

File: server/gitlab/gitlab.go [hunk 8]
```diff
@@ -20,6 +20,7 @@ var (
 	ErrPrivateResource = errors.New("private resource")
 )
 
+//go:generate mockgen -source=gitlab.go -destination=mocks/mock_gitlab.go
```
Comment (hanzei) [inline server/gitlab/gitlab.go:23]: Should this command be in `mocks` instead?

File: server/command_test.go [hunk 9]
```diff
@@ -363,3 +363,66 @@ func TestAddWebhookCommand(t *testing.T) {
...
+			defer mockCtrl.Finish()
+			mockedClient := mocks.NewMockGitlab(mockCtrl)
+
+			mockedClient.EXPECT().TriggerNewBuildPipeline(gomock.Any(), gomock.Any(), gomock.Any()).Return(test.pipeline, nil)
+			p.GitlabClient = mockedClient
+
+			api := &plugintest.API{}
+			pipeline := &gitLabAPI.Pipeline{}
+
+			api.On("TriggerNewBuildPipeline", mock.Anything, mock.Anything, mock.Anything).Return(pipeline)
```
Comment (hanzei) [inline server/command_test.go:421]: It would be nice to also assert the arguments used

File: server/command_test.go [hunk 10]
```diff
@@ -363,3 +363,66 @@ func TestAddWebhookCommand(t *testing.T) {
 		})
 	}
 }
+
+type buildCommandTest struct {
+	testName string
+	// scope      string
```
Comment (hanzei) [inline server/command_test.go:369]: Remove it not used?

File: server/api.go [hunk 11]
```diff
@@ -84,6 +84,8 @@ func (p *Plugin) ServeHTTP(c *plugin.Context, w http.ResponseWriter, r *http.Req
 		p.updateSettings(w, r)
 	case "/api/v1/user":
 		p.getGitlabUser(w, r)
+	case "/api/v1/build":
```
Comment (hanzei) [inline server/api.go:87]: Can you please elaborate why an API endpoint is needed?

File: server/gitlab/api.go [hunk 12]
```diff
@@ -412,3 +412,35 @@ func (g *gitlab) ResolveNamespaceAndProject(
 	}
 	return "", "", ErrNotFound
 }
+
+func (g *gitlab) TriggerNewBuildPipeline(user *UserInfo, repo interface{}, refToCommit *string) (*internGitlab.Pipeline, error) {
+	var err error
```
Comment (hanzei) [inline server/gitlab/api.go:417]: Nit pick: is this line needed?

File: server/gitlab/api.go [hunk 13]
```diff
@@ -412,3 +412,35 @@ func (g *gitlab) ResolveNamespaceAndProject(
 	}
 	return "", "", ErrNotFound
 }
+
+func (g *gitlab) TriggerNewBuildPipeline(user *UserInfo, repo interface{}, refToCommit *string) (*internGitlab.Pipeline, error) {
```
Comment (hanzei) [inline server/gitlab/api.go:416]: Can we pass the `refToCommit` directly instead of using a pointer?

Comment (hanzei) [inline server/gitlab/api.go:416]: I know that `client.Pipelines.CreatePipeline` accepts `repo` as `string` and `int`. Maybe we just stick with `string` for simplicity?

File: server/gitlab/api.go [hunk 14]
```diff
@@ -412,3 +412,35 @@ func (g *gitlab) ResolveNamespaceAndProject(
...
+		return nil, err
+	}
+
+	var (
+		pipeline       *internGitlab.Pipeline
+		ctx, ctxCancel = context.WithTimeout(context.Background(), DefaultRequestTimeout)
+	)
+
+	defer ctxCancel()
+	errGroup, _ := errgroup.WithContext(ctx)
```
Comment (hanzei) [inline server/gitlab/api.go:429]: Why is an `errgroup` needed for only one go routine?

File: server/plugin.go [hunk 15]
```diff
@@ -245,36 +245,41 @@ func (p *Plugin) CreateBotDMPost(userID, message, postType string) *model.AppErr
 }
 
 func (p *Plugin) PostToDo(info *gitlab.UserInfo) {
-	text, err := p.GetToDo(info)
+	hasTodo, text, err := p.GetToDo(info)
```
Comment (hanzei) [inline server/plugin.go:248]: Is this an unrelated change?

File: server/gitlab/api.go [hunk 16]
```diff
@@ -412,3 +412,35 @@ func (g *gitlab) ResolveNamespaceAndProject(
...
+	errGroup.Go(func() error {
+		p, response, gitError := client.Pipelines.CreatePipeline(repo, &internGitlab.CreatePipelineOptions{
```
Comment (hanzei) [inline server/gitlab/api.go:431]: Will this block until the pipeline is done?

File: server/gitlab/api.go [hunk 17]
```diff
@@ -412,3 +412,35 @@ func (g *gitlab) ResolveNamespaceAndProject(
...
+		if gitError != nil && response != nil && response.StatusCode != http.StatusNotFound {
+			return fmt.Errorf("failed to retrieve project by path: %w", gitError)
+		}
+		pipeline = p
+		return nil
+	})
+
+	if err = errGroup.Wait(); err != nil {
+		return nil, err
+	}
```
Comment (hanzei) [inline server/gitlab/api.go:444]: I would love to see some kind of report from the Pipeline in Mattermost.

**Review by hanzei — CHANGES_REQUESTED** [review]
This looks exiting :tada: 

There is still a lot of tune, but the base changes look good.

File: server/command.go (hunk 4 above)
Comment (tw-ayush) [reply server/command.go:646]: Oh my bad, seems like a blunder on my side

File: server/gitlab/api.go (hunk 17 above)
Comment (tw-ayush) [reply server/gitlab/api.go:444]: the pipeline struct return in case of success contains the details related to build 

`type Pipeline struct {
	ID             int             `json:"id"`
	Status         string          `json:"status"`
	Ref            string          `json:"ref"`
	SHA            string          `json:"sha"`
	BeforeSHA      string          `json:"before_sha"`
	Tag            bool            `json:"tag"`
	YamlErrors     string          `json:"yaml_errors"`
	User           *BasicUser      `json:"user"`
	UpdatedAt      *time.Time      `json:"updated_at"`
	CreatedAt      *time.Time      `json:"created_at"`
	StartedAt      *time.Time      `json:"started_at"`
	FinishedAt     *time.Time      `json:"finished_at"`
	CommittedAt    *time.Time      `json:"committed_at"`
	Duration       int             `json:"duration"`
	Coverage       string          `json:"coverage"`
	WebURL         string          `json:"web_url"`
	DetailedStatus *DetailedStatus `json:"detailed_status"`
}`

Comment (tw-ayush) [reply server/gitlab/api.go:444]: also inside the Detailed status there is an image as well

File: server/gitlab/api.go (hunk 16 above)
Comment (tw-ayush) [reply server/gitlab/api.go:431]: yes the code in lib looks blocking in nature.

File: server/gitlab/gitlab.go (hunk 8 above)
Comment (tw-ayush) [reply server/gitlab/gitlab.go:23]: no this command simply help any other user easily update the mock after any code change in interface as well as the build and test can automatically create the mock files/structs.

File: server/api.go (hunk 11 above)
Comment (tw-ayush) [reply server/api.go:87]: I assumed like all the other commands this can also be integrated in web?

File: server/command.go (hunk 7 above)
Comment (tw-ayush) [reply server/command.go]: seems like it, I've made the changes

File: server/command.go (hunk 6 above)
Comment (tw-ayush) [reply server/command.go:361]: How else this would be called other wise? if i don't add this here the tests won't run like they does for the other commands

**Comment by tw-ayush** [comment]
@hanzei please have a check now

File: server/api.go (hunk 11 above)
Comment (hanzei) [reply server/api.go:87]: Right now all slash commands are not supported via the Webapp. Let's remove it for now.

File: server/command.go (hunk 6 above)
Comment (hanzei) [reply server/command.go:361]: Doesn't L325 ff. already contain this code? Why does it need to be duplicated?

File: server/command.go [hunk 18]
```diff
@@ -600,7 +636,11 @@ func getAutocompleteData() *model.AutocompleteData {
 	webhookAdd.AddTextArgument("[Optional] token: Secret for webhook. Defaults to token used in plugin's settings.", "[token]", "")
 	webhook.AddCommand(webhookAdd)
 
-	gitlabCommand.AddCommand(webhook)
+	webhookBuild := model.NewAutocompleteData(commandBuild, "owner/[repo] reference", "trigger the pipeline")
```
Comment (hanzei) [inline server/command.go:639]: Maybe `build` instead of `webhookBuild`?

File: server/command.go [hunk 19]
```diff
@@ -600,7 +636,11 @@ func getAutocompleteData() *model.AutocompleteData {
...
+	webhookBuild.AddTextArgument("Group or Project path: includes user or group name with optional slash project name", "owner[/repo]", "")
+	webhookBuild.AddTextArgument("reference: reference/branch name for which the pipeline needs to be triggered", "reference", "")
+
+	gitlabCommand.AddCommand(webhookBuild)
```
Comment (hanzei) [inline server/command.go:643]: Let's order this alphabetically.

File: server/gitlab/api.go (hunk 17 above)
Comment (hanzei) [reply server/gitlab/api.go:431]: Would you be interested in also adding a success message to this PR?

**Review by hanzei — CHANGES_REQUESTED** [review]
We are getting there :+1:

File: server/command.go (hunk 6 above)
Comment (tw-ayush) [reply server/command.go:361]: Line 325 is part of commandAdd switch case

Comment (hanzei) [reply server/command.go:361]: Yes, but why does code need to get duplicated?

Comment (tw-ayush) [reply server/command.go:361]: Yes, you are right i can abstract it out in a function, i was just going with the rule of three
https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)

File: server/gitlab/api.go (hunk 17 above)
Comment (tw-ayush) [reply server/gitlab/api.go:431]: makes sense.

**Comment by tw-ayush** [comment]
> This PR has been automatically labelled "stale" because it hasn't had recent activity.
> A core team member will check in on the status of the PR to help with questions.
> Thank you for your contribution!
> 
> /cc @jasonblais @jfrerich @emilyacook

last 15 days were a festival season in INDIA, so I could not spare any time along with my day job to make the changes, I'll probably do it this week 😄

**Comment by jasonblais** [comment]
Thank you for the heads up @tw-ayush!

File: server/command.go [hunk 20]
```diff
@@ -600,8 +639,6 @@ func getAutocompleteData() *model.AutocompleteData {
 	webhookAdd.AddTextArgument("[Optional] token: Secret for webhook. Defaults to token used in plugin's settings.", "[token]", "")
 	webhook.AddCommand(webhookAdd)
 
-	gitlabCommand.AddCommand(webhook)
```
Comment (hanzei) [inline server/command.go:603]: Is this an unrelated change?

File: server/command.go [hunk 21]
```diff
@@ -341,6 +357,23 @@ func (p *Plugin) webhookCommand(parameters []string, info *gitlab.UserInfo) stri
...
+
+		pipeline, err := p.GitlabClient.TriggerNewBuildPipeline(info, project, ref)
+		if err != nil {
+			if strings.Contains(err.Error(), projectNotFoundError) {
+				return projectNotFoundMessage + project
+			}
+			return err.Error()
+		}
+
+		return pipeline.String()
```
Comment (hanzei) [inline server/command.go:375]: Does this code duplicate L234ff ?

**Review by hanzei — CHANGES_REQUESTED** [review]
One question, but looks good otherwise

File: server/command.go (hunk 21 above)
Comment (tw-ayush) [reply server/command.go:375]: yes it does, I tried to run it outside of webhook but that does not work for me, So i'll remove the L234 code and keep this one.

File: server/command.go (hunk 20 above)
Comment (tw-ayush) [reply server/command.go:603]: yup need to remove this change

**Review by hanzei — APPROVED** [review]
Code LGTM, but I haven't tested the feature

**Comment by tw-ayush** [comment]
> Code LGTM, but I haven't tested the feature

hey @hanzei thanks for Quick review and test updates

**Comment by jfrerich** [comment]
@iomodo, kind ping for review

**Review by iomodo — APPROVED** [review]
Looks great! Thanks @tw-ayush for the contribution!

**Comment by jfrerich** [comment]
@DHaussermann, ping for review and welcome back from the holiday break!

**Comment by tw-ayush** [comment]
> @tw-ayush or @hanzei I'm afraid I'm not familiar enough with GitLab pipelines to tests this. Can you please clarify what value from GitLab is passed in for `reference/branch name`? I have a pipeline setup on a GitLab project and it runs when I click `Run Pipeline` but I'm not sure what to pass into the slash command. The branch name only does not work and I don't know where to find the `reference`.
> 
> Can you provide some guidance and an example of what would be passed in after the project?
> 
> Also, it seems `build` has not been added to the help info for this new feature

you should be able to run the command by passing `<command> <project-name> <branch-name>`

**Comment by tw-ayush** [comment]
@hanzei this bring me back to my earlier doubt that we have used the create Pipeline Methods of SDK under the hood while it looks like a use case for trigger Pipeline APIs

**Comment by DHaussermann** [comment]
Thanks @tw-ayush that seems consistent with what I've been trying. I have a project named `pipelinetest` with a master branch so I tried `/gitlab build dhaussermann/pipelinetest master` But I get the response `Unknown action, please use /gitlab help to see all actions available.` Is there a sub-command that must be used after `build`?

Subscriptions to this same project are working. This branch seems to be properly deployed as I see `build` in the autocomplete. Also when firing it off via the GitLab UI, the test portion of my pipeline is failing but that does not seem related at all.

**Comment by tw-ayush** [comment]
> Thanks @tw-ayush that seems consistent with what I've been trying. I have a project named `pipelinetest` with a master branch so I tried `/gitlab build dhaussermann/pipelinetest master` But I get the response `Unknown action, please use /gitlab help to see all actions available.` Is there a sub-command that must be used after `build`?
> 
> Subscriptions to this same project are working. This branch seems to be properly deployed as I see `build` in the autocomplete. Also when firing it off via the GitLab UI, the test portion of my pipeline is failing but that does not seem related at all.

The tests also seems to be working fine.
I'm so far unable to set-up a sever and plugin to test these changes for myself, thus I only rely on unit/functional tests.
@hanzei can you help us out here?

**Comment by tw-ayush** [comment]
@DHaussermann can you please try running it under webhook command, seems like i've placed the execution call at a wrong place. Please try this once and let me know if that works.

**Comment by tw-ayush** [comment]
@DHaussermann did that work for you?

**Comment by DHaussermann** [comment]
@tw-ayush Sorry for the late reply.

1. Yes, it seems you're correct and this was put in the wrong place. Using the same project as above, when I run `/gitlab webhook build dhaussermann/pipelinetest master` it fires off my CI pipeline.

2. It does seem like there is an issue with the ephemeral post that gets created. Can you please see what happening here? This does not look like the intended behavior and it's showing text links to visual elements.
Screen Shot 2021-01-22 at 10 51 54 AM

**Comment by tw-ayush** [comment]
> @tw-ayush Sorry for the late reply.
> 
> 1. Yes, it seems you're correct and this was put in the wrong place. Using the same project as above, when I run `/gitlab webhook build dhaussermann/pipelinetest master` it fires off my CI pipeline.
> 2. It does seem like there is an issue with the ephemeral post that gets created. Can you please see what happening here? This does not look like the intended behavior and it's showing text links to visual elements.
>    Screen Shot 2021-01-22 at 10 51 54 AM

1. I'll try and move it to an appropriate end point. I may need to figure out how to test that end point.
2. Writing a to string end point should probably fix this. Can you tell me which are the expected fields in the response.

**Comment by emilyacook** [comment]
@DHaussermann gentle ping :)

**Comment by DHaussermann** [comment]
Hi @tw-ayush sorry for the late reply. 
I took a look at the post and all the links are reachable and look valid. 

Regarding what fields are expected in the response - We don't exactly have a matching use case to compare this to so I do not have a list of what fields we should include in the post. I'd say there's some room for discretion here. In my opinion, we don't need all the fields that are currently included. 
0/5 Maybe these details would be enough:
for the pipeline: `gitlab.Pipeline{ID:252742578, Status:"pending", Ref:"master",` 
For the user who triggered it: `Username:"dhaussermann"`
For the link: `The Url to the pipeline build details https://gitlab.com/dhaussermann/pipelinetest/-/pipelines/252742578`
@aaronrothschild any thoughts on this ☝️ ? You can see from the screen above the post that get's created when the pipline is triggered would be a bit messy and it would be difficult top format this all into an obvious and clear piece of information.    

Also @tw-ayush  please let me know when you have a chance to move the `build` command

**Comment by tw-ayush** [comment]
> Hi @tw-ayush sorry for the late reply.
> I took a look at the post and all the links are reachable and look valid.
> 
> Regarding what fields are expected in the response - We don't exactly have a matching use case to compare this to so I do not have a list of what fields we should include in the post. I'd say there's some room for discretion here. In my opinion, we don't need all the fields that are currently included.
> 0/5 Maybe these details would be enough:
> for the pipeline: `gitlab.Pipeline{ID:252742578, Status:"pending", Ref:"master",`
> For the user who triggered it: `Username:"dhaussermann"`
> For the link: `The Url to the pipeline build details https://gitlab.com/dhaussermann/pipelinetest/-/pipelines/252742578`
> @aaronrothschild any thoughts on this ☝️ ? You can see from the screen above the post that get's created when the pipline is triggered would be a bit messy and it would be difficult top format this all into an obvious and clear piece of information.
> 
> Also @tw-ayush please let me know when you have a chance to move the `build` command

thanks for the update.
Due to some personal commitments this is getting delayed. I'll update you soon on the changes.

**Comment by tw-ayush** [comment]
> Hi @tw-ayush sorry for the late reply.
> I took a look at the post and all the links are reachable and look valid.
> 
> Regarding what fields are expected in the response - We don't exactly have a matching use case to compare this to so I do not have a list of what fields we should include in the post. I'd say there's some room for discretion here. In my opinion, we don't need all the fields that are currently included.
> 0/5 Maybe these details would be enough:
> for the pipeline: `gitlab.Pipeline{ID:252742578, Status:"pending", Ref:"master",`
> For the user who triggered it: `Username:"dhaussermann"`
> For the link: `The Url to the pipeline build details https://gitlab.com/dhaussermann/pipelinetest/-/pipelines/252742578`
> @aaronrothschild any thoughts on this ☝️ ? You can see from the screen above the post that get's created when the pipline is triggered would be a bit messy and it would be difficult top format this all into an obvious and clear piece of information.
> 
> Also @tw-ayush please let me know when you have a chance to move the `build` command

thanks for the update.
Due to some personal commitments this is getting delayed. I'll update you soon on the changes.

**Comment by tw-ayush** [comment]
@DHaussermann i've moved the command to appropriate place.
@hanzei it won't be possible to keep the tests now, since the method call some configuration code. I believe that's the reason there are no tests written for ExecuteCommand method.

**Review by hanzei — APPROVED** [review]

**Review by iomodo — APPROVED** [review]

**Comment by hahmadia** [comment]
@DHaussermann Pinging for QA review. Thanks.

**Comment by DHaussermann** [comment]
Thanks @tw-ayush for the last commit. Confirmed the new functionality now works correctly with the `build` sub-command 👍 

Seems like what still needed here is to clean-up the response to be more user friendly. We should show the profile image properly and format some of this data to be more readable. @aaronrothschild  Do we need requirements on this? Seems like there is alot of info that could be useful to include in the post. Maybe the points outlined here https://github.com/mattermost/mattermost-plugin-gitlab/pull/224#issuecomment-774837633 would work?

**Comment by tw-ayush** [comment]
> /update-branch

done

**Comment by tw-ayush** [comment]
> Thanks @tw-ayush for the last commit. Confirmed the new functionality now works correctly with the `build` sub-command 👍
> 
> Seems like what still needed here is to clean-up the response to be more user friendly. We should show the profile image properly and format some of this data to be more readable. @aaronrothschild Do we need requirements on this? Seems like there is alot of info that could be useful to include in the post. Maybe the points outlined here [#224 (comment)](https://github.com/mattermost/mattermost-plugin-gitlab/pull/224#issuecomment-774837633) would work?

I've refined the response. Though I believe sending an image along with the response can be a bit tricky so for the time i've just sent the link to the avatar only.

**Comment by tw-ayush** [comment]
@DHaussermann gental reminder.

**Comment by DHaussermann** [comment]
Thanks for the last commit on this @tw-ayush. Sorry for the delay. 

I do see the improvements you've made on this 👍 . 

@aaronrothschild not sure what the best course is to close this PR off and get this feature in. 
0/5 but my thoughts are:
- If displaying the avatar is tricky, I would be completely fine if the link were removed from the post.   
- We may also want remove the name and active status

The response now looks like this. IF it's easy to remove more info to make the post simpler I would be in favor of it.
Screen Shot 2021-03-25 at 5 14 27 PM

**Comment by aaronrothschild** [comment]
> Thanks for the last commit on this @tw-ayush. Sorry for the delay.
> 
> I do see the improvements you've made on this 👍 .
> 
> @aaronrothschild not sure what the best course is to close this PR off and get this feature in.
> 0/5 but my thoughts are:
> 
> * If displaying the avatar is tricky, I would be completely fine if the link were removed from the post.
> * We may also want remove the name and active status
> 
> The response now looks like this. IF it's easy to remove more info to make the post simpler I would be in favor of it.
> Screen Shot 2021-03-25 at 5 14 27 PM

@abhijit-singh could you determine exactly which fields we should post and how they should be formatted as a Mattermost Message?  

we don't need the avatar URL, But should have: Pipeline, Status, Username, Ref:, and a link back to the pipeline page.

**Comment by maisnamrajusingh** [comment]
@hanzei im assigning this to @lufty-demansol

**Comment by lufty-demansol** [comment]
@tw-ayush Hey. I'm having problem locating your email. Can I please have permission to clone your PR to my repo and continue the work ?

**Comment by tw-ayush** [comment]
> @tw-ayush Hey. I'm having problem locating your email. Can I please have permission to clone your PR to my repo and continue the work ?

Hey @lufty-demansol please go ahead.

File: server/command.go (hunk 5 above)
Comment (lufty-demansol) [reply server/command.go]: seems to be fixed here: https://github.com/lufty-demansol/mattermost-plugin-gitlab/blob/master/server/command.go#L563

File: server/gitlab/api.go (hunk 13 above)
Comment (lufty-demansol) [reply server/gitlab/api.go]: Current code looks fine: https://github.com/lufty-demansol/mattermost-plugin-gitlab/blob/master/server/gitlab/api.go#L416

**Comment by lufty-demansol** [comment]
> Thanks for the last commit on this @tw-ayush. Sorry for the delay.
> 
> I do see the improvements you've made on this 👍 .
> 
> @aaronrothschild not sure what the best course is to close this PR off and get this feature in. 0/5 but my thoughts are:
> 
> * If displaying the avatar is tricky, I would be completely fine if the link were removed from the post.
> * We may also want remove the name and active status
> 
> The response now looks like this. IF it's easy to remove more info to make the post simpler I would be in favor of it. Screen Shot 2021-03-25 at 5 14 27 PM

@DHaussermann I continue the work in [this PR](https://github.com/mattermost/mattermost-plugin-gitlab/pull/280) and remove couple of response fields.

cc @maisnamrajusingh

**Comment by hanzei** [comment]
Closing as https://github.com/mattermost/mattermost-plugin-gitlab/issues/213 has been fixed by another PR
